prost-build/src/code_generator.rs

```diff
@@ -251,26 +251,35 @@ impl<'a> CodeGenerator<'a> {
 
     fn append_type_attributes(&mut self, fq_message_name: &str) {
         assert_eq!(b'.', fq_message_name.as_bytes()[0]);
-        // TODO: this clone is dirty, but expedious.
-        if let Some(attributes) = self.config.type_attributes.get(fq_message_name).cloned() {
-            self.push_indent();
-            self.buf.push_str(&attributes);
-            self.buf.push('\n');
+        let mut buf = String::new();
+        for attribute in self.config.type_attributes.get(fq_message_name) {
+            for _ in 0..self.depth {
+                buf.push_str("    ");
+            }
+            buf.push_str(&attribute);
+            buf.push('\n');
+        }
+        if !buf.is_empty() {
+            self.buf.push_str(&buf);
         }
     }
 
     fn append_field_attributes(&mut self, fq_message_name: &str, field_name: &str) {
         assert_eq!(b'.', fq_message_name.as_bytes()[0]);
-        // TODO: this clone is dirty, but expedious.
-        if let Some(attributes) = self
+        let mut buf = String::new();
```

Comment:
What is the benefit of this extra String? I don't see why you can't just push directly into the buf.

Reply:
The extra string is adressing the removed TODO.

It avoids the need for cloning the map (lifetime issue between `&self` for itering the map and `&mut self` for editing the buffer).

Reply:
Ah got it didn't catch that borrowing issue.

Reply:
I improved it to remove the local buffer, only the call to `self.push_ident()` was problematic.